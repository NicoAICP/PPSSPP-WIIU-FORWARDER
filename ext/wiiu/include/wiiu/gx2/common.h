#pragma once

#define GX2_SCAN_BUFFER_ALIGNMENT      0x1000
#define GX2_SHADER_ALIGNMENT           0x100
#define GX2_CONTEXT_STATE_ALIGNMENT    0x100
#define GX2_UNIFORM_BLOCK_ALIGNMENT    0x100
#define GX2_DISPLAY_LIST_ALIGNMENT     0x20
#define GX2_VERTEX_BUFFER_ALIGNMENT    0x40
#define GX2_INDEX_BUFFER_ALIGNMENT     0x20

#define GX2_ENABLE                     TRUE
#define GX2_DISABLE                    FALSE

#define GX2_TRUE                       TRUE
#define GX2_FALSE                      FALSE

#ifndef GX2_COMP_SEL
#define GX2_COMP_SEL(c0, c1, c2, c3) (((c0) << 24) | ((c1) << 16) | ((c2) << 8) | (c3))
#ifdef __cplusplus
#include <functional>
#endif
#include <stdio.h>

#define _x 0
#define _y 1
#define _z 2
#define _w 3
#define _r 0
#define _g 1
#define _b 2
#define _a 3
#define _0 4
#define _1 5
#endif


#define GX2_DISABLE_WRAPS
