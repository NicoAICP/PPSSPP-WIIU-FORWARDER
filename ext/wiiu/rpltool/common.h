#ifndef COMMON_H
#define COMMON_H

#include <ctype.h>
#include <stdint.h>
#include <stdio.h>

typedef int8_t s8;
typedef int16_t s16;
typedef int32_t s32;

typedef uint8_t u8;
typedef uint16_t u16;
typedef uint32_t u32;

#define GET_ENUM_NAME(prefix, name, val) \
   case prefix##name: \
      return #name;

#define EMIT_ENUM_TO_STR_FUNC(name, cases) \
   static inline const char *name##_to_str(name type) { \
      switch (type) { cases } \
      return "unknown"; \
   }

#define GET_FLAG_NAME(prefix, name, val) \
   case prefix##name: \
      pos += snprintf(buffer + pos, sizeof(buffer) - pos, "%s%s", #name, seperator); \
      flags &= ~(1u << i); \
      break;

#define EMIT_FLAGS_TO_STR_FUNC(name, cases) \
   static inline const char *name##_to_str(int flags, const char *seperator) { \
      int i; \
      static char buffer[32 * 32]; \
      int pos = 0; \
      for (i = 0; i < 32; i++) \
         if ((1u << i) & flags) \
            switch (1u << i) { cases } \
      if (flags) \
         snprintf(buffer + pos, sizeof(buffer) - pos, "0x%08X", flags); \
      else if (pos) \
         buffer[pos - strlen(seperator)] = '\0'; \
      else \
         return "NULL"; \
      return buffer; \
   }

#define MAKE_MAGIC(a, b, c, d) (((a) << 24) | ((b) << 16) | ((c) << 8) | ((d) << 0))

#define log_var(X) printf("%-35s: 0x%0*X (%u)\n", #X, sizeof(X) * 2, (u32)(X), (u32)(X))
#define log_svar(X) printf("%-35s: 0x%0*X (%i)\n", #X, sizeof(X) * 2, (u32)(X), (s32)(X))
#define log_array(X) debug_array(X, sizeof(X), #X)
#define log_enum(type, X) printf("%-35s: 0x%X (%s)\n", #X, (X), type##_to_str(X))
#define log_flags(type, X) printf("%-35s: 0x%08X (%s)\n", #X, (X), type##_to_str(X, " | "))
#define log_magic(X) \
   printf("%-35s: '%c''%c''%c''%c' (0x%08X)\n", #X, (u32)(X) >> 24, (u32)(X) >> 16, (u32)(X) >> 8, (u32)(X), (u32)(X))
#define log_carray(X) debug_char_array(X, sizeof(X), #X)
#define log_str(X) printf("%-35s: \"%s\"\n", #X, (char *)(X))
#define log_off_str(X, data) printf("%-35s: 0x%08X \"%s\"\n", #X, (X), (char *)data + (X))
#define log_off_strs(X, data) \
   do { \
      printf("%-35s: 0x%08X ", #X, (X)); \
      const char *ptr = (char *)data + (X); \
      while (*ptr) { \
         ptr += printf("\"%s\" ", ptr) - 2; \
      } \
      printf("\n"); \
   } while (0)

static inline void dump_array(uint32_t *src, int size);
static inline void debug_array(uint32_t *src, int size, const char *name) {
   printf("%-35s:", name);
   dump_array(src, size);
}
static inline void debug_char_array(uint8_t *src, int size, const char *name) {
   int i;
   printf("%-35s: ", name);
   //   for (i = 0; i < size; i++)
   //      printf("%-2c ", isalnum(src[i])?src[i]: '?');
   //   printf("\n%-35s: ", "");
   for (i = 0; i < size; i++)
      printf("%02X ", src[i]);
   printf("\n");
}

static inline void dump_array(uint32_t *src, int size) {
   int i;
   size = (size / 4) - 1;
   printf("\n{");
   for (i = 0; i < size; i++) {
      if (!(i & 3))
         printf("\n   ");

      printf("0x%08X, ", src[i]);
   }

   if (!(i & 3))
      printf("\n   ");

   printf("0x%08X\n}\n", src[i]);
}

static inline u32 align_up(u32 x, u32 align) { return (x + (align - 1)) & ~(align - 1); }
static inline u32 align_down(u32 x, u32 align) { return x & ~(align - 1); }

#endif // COMMON_H
