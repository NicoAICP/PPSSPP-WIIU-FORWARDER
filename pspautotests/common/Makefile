TARGET_LIB = libcommon.a
EXTRA_TARGETS = libcommon_kernel.a
OBJS = common.o vram.o sysmem-imports.o
KERNEL_OBJS = common_kernel.o vram.o sysmem-imports.o

INCDIR = 
CFLAGS = -O0 -G 0 -g -Wall
CXXFLAGS = $(CFLAGS) -fno-exceptions -fno-rtti
ASFLAGS = $(CFLAGS)

LIBDIR =
LDFLAGS = -G0

PSPSDK=$(shell psp-config --pspsdk-path)
include $(PSPSDK)/lib/build.mak

%.o: %.S
	$(AS) $(ASFLAGS) -c -o $@ $<

common_kernel.o: common.c
	$(CC) $(CFLAGS) -DCOMMON_KERNEL=1 -c -o $@ $<

libcommon_kernel.a: $(KERNEL_OBJS)
	$(AR) cru $@ $(KERNEL_OBJS)
	$(RANLIB) $@
