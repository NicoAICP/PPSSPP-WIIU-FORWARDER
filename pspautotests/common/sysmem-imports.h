extern int sceKernelSetCompiledSdkVersion(int ver);
extern int sceKernelSetCompiledSdkVersion370(int ver);
extern int sceKernelSetCompiledSdkVersion380_390(int ver);
extern int sceKernelSetCompiledSdkVersion395(int ver);
extern int sceKernelSetCompiledSdkVersion401_402(int ver);
extern int sceKernelSetCompiledSdkVersion500_505(int ver);
extern int sceKernelSetCompiledSdkVersion507(int ver);
extern int sceKernelSetCompiledSdkVersion600_602(int ver);
extern int sceKernelSetCompiledSdkVersion603_605(int ver);
extern int sceKernelSetCompiledSdkVersion606(int ver);
extern int sceKernelGetCompiledSdkVersion();