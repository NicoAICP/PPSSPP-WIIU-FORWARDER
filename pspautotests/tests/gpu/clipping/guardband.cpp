#include <common.h>
#include <pspdisplay.h>
#include <pspgu.h>
#include <pspkernel.h>
#include <pspthreadman.h>
#include "../commands/commands.h"

extern "C" int sceDmacMemcpy(void *dest, const void *source, unsigned int size);

typedef struct {
	u32 color;
	float x, y, z;
} VertexColorF32;

enum TriangleType {
	TRIANGLE_ALL = -1,
	TRIANGLE_NORMAL = 0,
	TRIANGLE_AT_EDGE = 1,

	TRIANGLE_OUT_NEG_X = 2,
	TRIANGLE_OUT_NEG_Y = 3,
	TRIANGLE_OUT_NEG_Z = 4,

	TRIANGLE_OUT_POS_X = 5,
	TRIANGLE_OUT_POS_Y = 6,
	TRIANGLE_OUT_POS_Z = 7,
};

// +---------------------+  4096x4096 outer box
// |                     |  480x272 inner box
// |      +-------+      |
// |      |       |      |  Outer box at (-1808, -1912) - (2288, 2184)
// |      +-------+      |  Inner box at (0, 0) - (479, 271)
// |                     |
// +---------------------+
static __attribute__((aligned(16))) VertexColorF32 triangles[][3] = {
	// TRIANGLE_NORMAL
	{
		{0xFFFFFFFF, 0.0f, 0.0f, 0.0f},
		{0xFFFFFFFF, 0.0f, 272.0f, 0.5f},
		{0xFFFFFFFF, 480.0f, 272.0f, 1.0f},
	},
	// TRIANGLE_AT_EDGE
	{
		{0xFFFFFFFF, -1808.0f, -1912.0f, 0.0f},
		{0xFFFFFFFF, 0.0f, 2184.0f, 0.5f},
		{0xFFFFFFFF, 2288.0f, 2184.0f, 1.0f},
	},

	// TRIANGLE_OUT_NEG_X
	{
		{0xFFFFFFFF, -1809.0f, -1912.0f, 0.0f},
		{0xFFFFFFFF, 0.0f, 2184.0f, 0.5f},
		{0xFFFFFFFF, 2288.0f, 2184.0f, 1.0f},
	},
	// TRIANGLE_OUT_NEG_Y
	{
		{0xFFFFFFFF, -1808.0f, -1913.0f, 0.0f},
		{0xFFFFFFFF, 0.0f, 2184.0f, 0.5f},
		{0xFFFFFFFF, 2288.0f, 2184.0f, 1.0f},
	},
	// TRIANGLE_OUT_NEG_Z
	{
		{0xFFFFFFFF, -1808.0f, -1912.0f, -0.1f},
		{0xFFFFFFFF, 0.0f, 2184.0f, 0.5f},
		{0xFFFFFFFF, 2288.0f, 2184.0f, 1.0f},
	},

	// TRIANGLE_OUT_POS_X
	{
		{0xFFFFFFFF, 2289.0f, -1912.0f, 0.0f},
		{0xFFFFFFFF, -1808.0f, -1912.0f, 0.5f},
		{0xFFFFFFFF, 2289.0f, 2184.0f, 1.0f},
	},
	// TRIANGLE_OUT_POS_Y
	{
		{0xFFFFFFFF, 2288.0f, -1912.0f, 0.0f},
		{0xFFFFFFFF, -1808.0f, -1912.0f, 0.5f},
		{0xFFFFFFFF, 2288.0f, 2185.0f, 1.0f},
	},
	// TRIANGLE_OUT_POS_Z
	{
		{0xFFFFFFFF, -1808.0f, -1912.0f, 0.0f},
		{0xFFFFFFFF, 0.0f, 2184.0f, 0.5f},
		{0xFFFFFFFF, 2288.0f, 2184.0f, 1.1f},
	},
};

static u8 *fbp0 = 0;
static u8 *dbp0 = fbp0 + 512 * 272 * sizeof(u32);

static u32 *copybuf = NULL;
static unsigned int __attribute__((aligned(16))) list[32768];

inline VertexColorF32 makeVertex32(u32 c, float x, float y, float z) {
	VertexColorF32 v;
	v.color = c;
	v.x = x;
	v.y = y;
	v.z = z;
	return v;
}

void resetBuffer() {
	memset(copybuf, 0x44, 512 * 272 * 4);
	sceKernelDcacheWritebackInvalidateAll();
	sceDmacMemcpy(sceGeEdramGetAddr(), copybuf, 512 * 272 * 4);
	sceKernelDcacheWritebackInvalidateAll();
}

void displayBuffer(const char *reason) {
	sceKernelDcacheWritebackInvalidateAll();
	sceDmacMemcpy(copybuf, sceGeEdramGetAddr(), 512 * 272 * 4);
	sceKernelDcacheWritebackInvalidateAll();
	const u32 *buf = copybuf;

	bool found = false;
	for (int y = 0; y < 272 && !found; ++y) {
		for (int x = 0; x < 480; ++x) {
			if ((buf[y * 512 + x] & 0x00FFFFFF) == 0x00FFFFFF) {
				found = true;
				break;
			}
		}
	}
	checkpoint("%s: DRAW=%d", reason, found ? 1 : 0);
}

void drawBoxCommands(TriangleType tri) {
	// Clearing cache is fun.  Let's do it all the time.
	sceKernelDcacheWritebackInvalidateAll();
	if (tri >= 0) {
		sceGuDrawArray(GU_TRIANGLES, GU_COLOR_8888 | GU_VERTEX_32BITF | GU_TRANSFORM_3D, 3, NULL, triangles[tri]);
	} else {
		sceGuDrawArray(GU_TRIANGLES, GU_COLOR_8888 | GU_VERTEX_32BITF | GU_TRANSFORM_3D, sizeof(triangles) / sizeof(triangles[0][0]), NULL, triangles[0]);
	}
	sceDisplaySetFrameBuf(sceGeEdramGetAddr(), 512, GU_PSM_8888, PSP_DISPLAY_SETBUF_IMMEDIATE);
}

void init() { 
	copybuf = new u32[512 * 272];

	sceGuInit();
	sceGuStart(GU_DIRECT, list);
	sceGuDrawBuffer(GU_PSM_8888, fbp0, 512);
	sceGuDispBuffer(480, 272, fbp0, 512);
	sceGuDepthBuffer(dbp0, 512);
	sceGuOffset(2048 - (480 / 2), 2048 - (272 / 2));
	sceGuViewport(2048, 2048, 480, 272);
	sceGuDepthRange(0, 65535);
	sceGuDepthMask(0);
	sceGuScissor(0, 0, 480, 272);
	sceGuEnable(GU_SCISSOR_TEST);
	sceGuFrontFace(GU_CW);
	sceGuShadeModel(GU_SMOOTH);
	sceGuDisable(GU_TEXTURE_2D);
	sceGuDisable(GU_BLEND);

	sceGuSendCommandf(GE_CMD_VIEWPORTZ1, 32767.5);
	sceGuSendCommandf(GE_CMD_VIEWPORTZ2, 32767.5);
	sceGuSendCommandi(GE_CMD_MINZ, 0);
	sceGuSendCommandi(GE_CMD_MAXZ, 65535);

	ScePspFMatrix4 ones = {
		{1, 0, 0, 0},
		{0, 1, 0, 0},
		{0, 0, 1, 0},
		{0, 0, 0, 1},
	};

	ScePspFMatrix4 ortho = {
		{2.0f / 480.0f, 0, 0, 0},
		{0, -2.0f / 272.0f, 0, 0},
		{0, 0, 2.0f, 0},
		{-1.0f, 1.0f, -1.0f, 1.0f},
	};

	sceGuSetMatrix(GU_MODEL, &ones);
	sceGuSetMatrix(GU_VIEW, &ones);
	sceGuSetMatrix(GU_PROJECTION, &ortho);

	sceGuFinish();
	sceGuSync(0, 0);
 
	sceDisplayWaitVblankStart();
	sceGuDisplay(1);

	memset(copybuf, 0x44, 512 * 272 * 4);
	sceKernelDcacheWritebackInvalidateAll();
	sceDmacMemcpy(sceGeEdramGetAddr(), copybuf, 512 * 272 * 4);
	sceKernelDcacheWritebackInvalidateAll();

	displayBuffer("Initial");
}

void testTriangle(const char *title, TriangleType tri, bool clip = true) {
	resetBuffer();

	sceGuStart(GU_DIRECT, list);

	if (clip) {
		sceGuDisable(GU_CLIP_PLANES);
	} else {
		sceGuEnable(GU_CLIP_PLANES);
	}
	
	sceGuEnable(GU_DEPTH_TEST);
	sceGuDepthFunc(GU_ALWAYS);

	drawBoxCommands(tri);

	sceGuFinish();
	sceGuSync(GU_SYNC_WAIT, GU_SYNC_WHAT_DONE);
	sceDisplayWaitVblank();

	displayBuffer(title);
}

extern "C" int main(int argc, char *argv[]) {
	init();

	sceDisplaySetFrameBuf(sceGeEdramGetAddr(), 512, GU_PSM_8888, PSP_DISPLAY_SETBUF_IMMEDIATE);
	sceDisplaySetMode(0, 480, 272);

	testTriangle("  Normal, clipped", TRIANGLE_NORMAL, true);
	testTriangle("  Normal, unclipped", TRIANGLE_NORMAL, false);
	testTriangle("  Flat at edge, clipped", TRIANGLE_AT_EDGE, true);
	testTriangle("  Flat at edge, unclipped", TRIANGLE_AT_EDGE, false);

	checkpointNext("Flat out negative");
	testTriangle("  Flat out negative X, clipped", TRIANGLE_OUT_NEG_X, true);
	testTriangle("  Flat out negative X, unclipped", TRIANGLE_OUT_NEG_X, false);
	testTriangle("  Flat out negative Y, clipped", TRIANGLE_OUT_NEG_Y, true);
	testTriangle("  Flat out negative Y, unclipped", TRIANGLE_OUT_NEG_Y, false);
	testTriangle("  Flat out negative Z, clipped", TRIANGLE_OUT_NEG_Z, true);
	testTriangle("  Flat out negative Z, unclipped", TRIANGLE_OUT_NEG_Z, false);

	checkpointNext("Flat out positive");
	testTriangle("  Flat out positive X, clipped", TRIANGLE_OUT_POS_X, true);
	testTriangle("  Flat out positive X, unclipped", TRIANGLE_OUT_POS_X, false);
	testTriangle("  Flat out positive Y, clipped", TRIANGLE_OUT_POS_Y, true);
	testTriangle("  Flat out positive Y, unclipped", TRIANGLE_OUT_POS_Y, false);
	testTriangle("  Flat out positive Z, clipped", TRIANGLE_OUT_POS_Z, true);
	testTriangle("  Flat out positive Z, unclipped", TRIANGLE_OUT_POS_Z, false);

	checkpointNext("All (entire draw?)");
	testTriangle("  All, clipped", TRIANGLE_ALL, true);
	testTriangle("  All, unclipped", TRIANGLE_ALL, false);

	sceGuTerm();

	return 0;
}
