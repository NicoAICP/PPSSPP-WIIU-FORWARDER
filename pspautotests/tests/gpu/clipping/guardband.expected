[x] Initial: DRAW=0
[r]   Normal, clipped: DRAW=1
[r]   Normal, unclipped: DRAW=1
[r]   Flat at edge, clipped: DRAW=1
[r]   Flat at edge, unclipped: DRAW=1

[x] Flat out negative
[r]   Flat out negative X, clipped: DRAW=0
[r]   Flat out negative X, unclipped: DRAW=0
[r]   Flat out negative Y, clipped: DRAW=0
[r]   Flat out negative Y, unclipped: DRAW=0
[r]   Flat out negative Z, clipped: DRAW=0
[r]   Flat out negative Z, unclipped: DRAW=1

[x] Flat out positive
[r]   Flat out positive X, clipped: DRAW=0
[r]   Flat out positive X, unclipped: DRAW=0
[r]   Flat out positive Y, clipped: DRAW=0
[r]   Flat out positive Y, unclipped: DRAW=0
[r]   Flat out positive Z, clipped: DRAW=0
[r]   Flat out positive Z, unclipped: DRAW=1

[x] All (entire draw?)
[r]   All, clipped: DRAW=1
[r]   All, unclipped: DRAW=1
